File: Leetcode/0218-TheSkylineProblem/java/Solution3.java
public class Solution3 {
    public List<List<Integer>> getSkyline(int[][] buildings) {
        HashSet<Integer> edgeSet = new HashSet<>();
        for (int[] b : buildings) {
            int left = b[0];
            int right = b[1];
            edgeSet.add(left);
            edgeSet.add(right);
        }
        List<Integer> edges = new ArrayList<Integer>(edgeSet);
        Collections.sort(edges);

        List<List<Integer>> answer = new ArrayList();
        for (int edge : edges) {
            int maxHeight = 0;
            for (int[] b : buildings) {
                int left = b[0];
                int right = b[1];
                int height = b[2];
                // If current building intersects with the sweep line, update maxHeight
                if (left <= edge && edge < right) {
                    maxHeight = Math.max(maxHeight, height);
                }
            }
            // If [position, maxHeight] is the first key point,
            // or the height has changed,
            // add it to the answer
            if (answer.isEmpty() || answer.get(answer.size() - 1).get(1) != maxHeight) {
                answer.add(Arrays.asList(edge, maxHeight));
            }
        }
        return answer;
    }
}

File: Leetcode/0042-TrappingRainWater/go/solution3.go
func trap(height []int) int {
    n := len(height)
    lmax, rmax := 0, 0
    left, right := 0, n-1
    res := 0
    for left <= right {
        lmax = max(lmax, height[left])
        rmax = max(rmax, height[right])
        if lmax < rmax {
            res += lmax-height[left]
            left++
        } else {
            res += rmax-height[right]
            right--
        }
    }
    return res
}